### ISSUE TITLE
Renaming generic struct field

### ISSUE THREAD

**Issue by thatdevsherry** [tafia/quick-xml#495]
Hi, thank you very much for working on this crate and keeping it alive :)

I've just started using rust for a hobby project and got stuck on a problem. I've looked at existing github issues but I couldn't find anything similar. If this is a duplicate, I'd appreciate if you can point me to the relevant issue.

## Problem statement
I have a struct that accepts a generic type and has a field of that type.

```rust
struct Parent<T: SomeTrait> {
    child: T,
}
```

The `Parent` struct can have any type in its `child` field provided it implements `SomeTrait`.

When I serialize `Parent`, I end up with a `<child>` tag, which is expected because `Parent` has the field `child`. However, I want to change the field name on serialization so I get the name of the struct that was set in the `child` field.

I looked at `#[serde(rename = "$value")]` but I believe that is intended for deserialization, because on serialization it replaces `<child>` with `<$value>`.

## Full code example

```rust
use quick_xml;
use serde::{Deserialize, Serialize};

trait Young {}

#[derive(Debug, Serialize, Deserialize)]
struct Parent<T: Young> {
    child: T,
}

#[derive(Debug, Serialize, Deserialize)]
struct Child1 {}
impl Young for Child1 {}

#[derive(Debug, Serialize, Deserialize)]
struct Child2 {
    pub some_attribute: String,
}
impl Young for Child2 {}

fn main() {
    let parent_of_child_1 = Parent { child: Child1 {} };
    let parent_of_child_2 = Parent {
        child: Child2 {
            some_attribute: "some_attr".to_string(),
        },
    };

    let parent_1_serialized = quick_xml::se::to_string(&parent_of_child_1).unwrap();
    println!("Parent of child 1 serialized: {:?}", parent_1_serialized); // Parent of child 1 serialized: "<Parent><child/></Parent>"

    let parent_2_serialized = quick_xml::se::to_string(&parent_of_child_2).unwrap();
    println!("Parent of child 2 serialized: {:?}", parent_2_serialized); // Parent of child 2 serialized: "<Parent><child some_attribute=\"some_attr\"/></Parent>"
}
```

How would I dynamically set the name of tag inside `<Parent>` tag?

Thank you

**Comment by Mingun** [comment]
If I understand your correctly, you want to replace `<child>` in the XML by name of the type:
```xml
<Parent>
  <Child1/>
</Parent>
```
and
```xml
<Parent>
  <Child2 some_attribute=\"some_attr\"/>
</Parent>
```
That is impossible to achieve without (semi-)manual implementation of Serialize and Deserialize traits. Probably you want to use an enum as a type for a `child` field:
```rust
enum Children {
  Child1(Child1),
  Child2(Child2),
}
struct Parent {
    child: Children,
}
```
Then it would be serialized to and deserialized from XMLs shown above.

If you really want to use dynamic names, you should implement (de)serialization of `Parent` struct manually:
```rust
trait Young {
  fn type_name() -> &'static str;
}

struct Parent<T: Young> {
  child: T,
}
impl<T: Serialize + Young> Serialize for Parent<T> {...}
impl<'de, T: Deserialize<'de> + Young> Deserialize<'de> for Parent<T> {...}
```
One possible implementation could be found [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=60869f0457af655042ecec5888c40593). Deserialization is not implemented optimally, I left that to you as an exercise.

**Comment by thatdevsherry** [comment]
Thank you so much for providing an example and even going ahead and sharing the rust playground link!

I had a feeling I would need to implement Serialize/Deserialize to some extent.

The solution I've come up with is to implement `Serialize` (semi-implement, just using the key from trait as you have written), and using `#[serde(rename = "$value")]` to handle struct name for deserialization as my requirement doesn't involve parsing an unknown struct.

Thank you for helping me out. We can close this issue :)

**Comment by Mingun** [comment]
> I looked at `#[serde(rename = "$value")]` but I believe that is intended for deserialization, because on serialization it replaces `<child>` with `<$value>`.

Different interpretation of a special name `$value` is just a bug, that will be fixed after merging #490. In any case, do not rely on type names -- they will never be used except one specific case. I plan to write a documentation about mapping Rust to XML in near future.

The special name `$value` also will be splitted at `#text` and `#any` (see corresponding commit in #490). In your case I do not see any reasons to use that special names.

**Comment by thatdevsherry** [comment]
I see, will proceed to implement deserialize or just use enums then.

Thank you
